src/main/java/io/divolte/server/filesinks/gcs/DynamicDelegatingOutputStream.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2017 GoDataDriven B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.divolte.server.filesinks.gcs;
+
+import java.io.IOException;
+import java.io.OutputStream;
+import java.util.Objects;
+
+import javax.annotation.ParametersAreNonnullByDefault;
+
+import com.google.common.base.Preconditions;
+
+/*
+ * OutputStream wrapper that allows to dynamically replace the wrapped stream.
+ */
+@ParametersAreNonnullByDefault
+public class DynamicDelegatingOutputStream extends OutputStream {
+    private OutputStream wrapped;
+
+    public DynamicDelegatingOutputStream() {
+        wrapped = null;
+    }
+
+    public void detachDelegate() throws IOException {
+        wrapped.flush();
```

Comment:
I added a state check that `wrapped == null` on attach and `wrapped != null` on detach. All invocations are changed to be in a `try-finally`, where detach happens in the `finally` block.